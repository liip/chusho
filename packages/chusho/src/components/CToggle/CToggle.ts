import {
  ref,
  provide,
  computed,
  InjectionKey,
  defineComponent,
  createElement,
  watchEffect,
  Ref,
} from '@vue/composition-api';

export const ToggleSymbol: InjectionKey<object> = Symbol();

export interface UseToggle {
  open: Readonly<Ref<boolean>>;
  toggle: Function;
}

interface ToggleProps {
  open: boolean;
}

export default defineComponent<ToggleProps>({
  name: 'CToggle',

  model: {
    prop: 'open',
    event: 'toggle',
  },

  props: {
    /**
     * Optionally bind the Toggle state with the parent component.
     */
    open: {
      type: Boolean,
      default: undefined,
    },
  },

  setup(props, { slots, emit }) {
    const open = ref(false);

    function toggle(): void {
      open.value = !open.value;
      // Update potential parent v-model value
      emit('toggle', open.value);
    }

    // Provide api to sub-components
    const api: UseToggle = {
      open: computed(() => open.value),
      toggle,
    };
    provide(ToggleSymbol, api);

    // Watch potential parent v-model value changes and update state accordingly
    watchEffect(() => {
      if (typeof props.open === 'boolean') {
        open.value = props.open;
      }
    });

    return () => createElement('div', slots.default && slots.default());
  },
});
