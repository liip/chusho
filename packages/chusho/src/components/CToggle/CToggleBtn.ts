import { CBtn } from '../CBtn';
import { VNodeData } from 'vue/types/umd';
import { inject, defineComponent, createElement } from '@vue/composition-api';
import { ToggleSymbol } from './CToggle';
import { UseToggle } from './CToggle';

export default defineComponent({
  name: 'CToggleBtn',

  inheritAttrs: false,

  setup(props, { attrs, slots, listeners }) {
    const toggle = inject(ToggleSymbol) as UseToggle;

    return () => {
      const componentData: VNodeData = {
        attrs: {
          ...attrs,
          'aria-expanded': `${toggle.open.value}`,
          'aria-controls': `chusho-toggle-${toggle.uuid}`,
        },
        on: {
          ...listeners,
          click: () => {
            toggle.toggle();
          },
        },
      };
      return createElement(CBtn, componentData, slots.default());
    };
  },
});
