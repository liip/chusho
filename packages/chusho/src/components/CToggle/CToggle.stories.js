import CToggle from './CToggle';
import CToggleBtn from './CToggleBtn';
import CToggleContent from './CToggleContent';

export default {
  title: 'Components|Toggle',
  component: CToggle,
  subcomponents: { CToggleBtn, CToggleContent },
  parameters: {
    componentSubtitle: 'Conditionnaly display some content.',
    options: {
      componentConfig: [
        {
          name: 'transition',
          type: { summary: 'object' },
          description:
            'Apply a common transition to all Toggles. The object can contain any Vue built-in [transition component props](https://vuejs.org/v2/api/#transition).\n\nFor example: `{ name: "fade", mode: "out-in" }`',
        },
      ],
    },
  },
};

export const Default = () => ({
  components: { CToggle, CToggleBtn, CToggleContent },
  template: `<CToggle>
    <CToggleBtn variant="medium default">Toggle</CToggleBtn>
    <CToggleContent class="bg-gray-200 px-4 py-3 rounded mt-2">
      Lorem ipsum dolor sit amet consectetur adipisicing elit. Laboriosam in, iste id nobis dolor excepturi dolore expedita vero quae. Nobis fuga cupiditate suscipit blanditiis, aliquid minima harum molestias pariatur tempora ab, libero quo maiores sapiente doloribus nihil commodi eaque accusantium praesentium! Nobis natus qui voluptate inventore molestias quisquam, consequuntur harum?
    </CToggleContent>
  </CToggle>`,
});

export const Controlled = () => ({
  components: { CToggle, CToggleBtn, CToggleContent },
  data() {
    return {
      toggleOpen: true,
    };
  },
  template: `<CToggle v-model="toggleOpen">
    <CToggleBtn variant="medium default">{{ toggleOpen ? 'Close' : 'Open' }}</CToggleBtn>
    <CToggleContent class="bg-gray-200 px-4 py-3 rounded mt-2">
      Lorem ipsum dolor sit amet consectetur adipisicing elit. Laboriosam in, iste id nobis dolor excepturi dolore expedita vero quae. Nobis fuga cupiditate suscipit blanditiis, aliquid minima harum molestias pariatur tempora ab, libero quo maiores sapiente doloribus nihil commodi eaque accusantium praesentium! Nobis natus qui voluptate inventore molestias quisquam, consequuntur harum?
    </CToggleContent>
  </CToggle>`,
});

Controlled.story = {
  parameters: {
    docs: {
      storyDescription:
        'You can control the Toggle status with the `v-model` directive, for example to make it open by default or to programatically change its state.',
    },
  },
};
