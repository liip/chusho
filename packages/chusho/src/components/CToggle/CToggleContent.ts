import {
  inject,
  h,
  defineComponent,
  PropType,
  BaseTransitionProps,
  Transition,
  mergeProps,
} from 'vue';

import { DollarChusho } from '../../types';
import { isPlainObject } from '../../utils/objects';
import { generateConfigClass } from '../../utils/components';
import componentMixin from '../shared';
import { ToggleSymbol } from './CToggle';
import { UseToggle } from './CToggle';

export default defineComponent({
  name: 'CToggleContent',

  mixins: [componentMixin],

  inheritAttrs: false,

  props: {
    /**
     * The object can contain any Vue built-in [transition component props](https://v3.vuejs.org/api/built-in-components.html#transition).
     *
     * For example: `{ name: "fade", mode: "out-in" }`.
     *
     * If you defined a default transition in the config and want to disable it, use `false`.
     */
    transition: {
      type: [Object, Boolean] as PropType<BaseTransitionProps | false>,
      default: null,
    },
  },

  setup(props, { slots, attrs }) {
    const toggle = inject(ToggleSymbol) as UseToggle;

    function renderContent() {
      const toggleConfig = inject<DollarChusho | null>('$chusho', null)?.options
        ?.components?.toggleContent;
      const elementProps: Record<string, unknown> = {
        id: toggle.uuid,
        ...generateConfigClass(toggleConfig?.class, props),
      };

      if (toggle.open.value) {
        return h('div', mergeProps(attrs, elementProps), slots);
      }
      return null;
    }

    return () => {
      const toggleConfig = inject<DollarChusho | null>('$chusho', null)?.options
        ?.components?.toggleContent;
      let transitionProps: BaseTransitionProps | null = null;

      if (isPlainObject(props.transition)) {
        transitionProps = props.transition;
      } else if (props.transition !== false && toggleConfig?.transition) {
        transitionProps = toggleConfig.transition;
      }

      return transitionProps
        ? h(Transition, transitionProps, renderContent)
        : renderContent();
    };
  },
});
