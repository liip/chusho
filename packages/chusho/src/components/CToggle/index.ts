import CToggle from './CToggle';
import CToggleBtn from './CToggleBtn';
import CToggleContent from './CToggleContent';

export { CToggle, CToggleBtn, CToggleContent };
