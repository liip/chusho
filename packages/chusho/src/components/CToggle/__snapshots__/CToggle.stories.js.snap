// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Components|Toggle Controlled 1`] = `
<div>
  <button
    aria-expanded="false"
    class="inline-block py-2 px-4 bg-blue-200 text-blue-900 rounded"
    type="button"
    variant="medium default"
  >
    Close
  </button>
   
  <!---->
</div>
`;

exports[`Storyshots Components|Toggle Default 1`] = `
<div>
  <button
    aria-expanded="false"
    class="inline-block py-2 px-4 bg-blue-200 text-blue-900 rounded"
    type="button"
    variant="medium default"
  >
    Toggle
  </button>
   
  <!---->
</div>
`;
